targets/csharp-unityv2/source/Shared/Internal/PlayFabHttp/PlayFabHTTP.cs

```diff
@@ -0,0 +1,337 @@
+            };
+        }
+
+        protected internal static PlayFabError GeneratePlayFabError(string json, object customData = null)
+        {
+            //deserialize the error
+            var errorDict = JsonWrapper.DeserializeObject<JsonObject>(json, PlayFabUtil.ApiSerializerStrategy);
+
+            Dictionary<string, List<string>> errorDetails = null;
+            if (errorDict.ContainsKey("errorDetails"))
+            {
+                var ed =
+                    JsonWrapper.DeserializeObject<Dictionary<string, List<string>>>(
+                        errorDict["errorDetails"].ToString());
+                errorDetails = ed;
+            }
+            //create new error object
+            return new PlayFabError
+            {
+                HttpCode = errorDict.ContainsKey("code") ? Convert.ToInt32(errorDict["code"]) : 400,
+                HttpStatus = errorDict.ContainsKey("status")
+                    ? (string)errorDict["status"]
+                    : "BadRequest",
+                Error = errorDict.ContainsKey("errorCode")
+                    ? (PlayFabErrorCode)Convert.ToInt32(errorDict["errorCode"])
+                    : PlayFabErrorCode.ServiceUnavailable,
+                ErrorMessage = errorDict.ContainsKey("errorMessage")
+                    ? (string)errorDict["errorMessage"]
+                    : string.Empty,
+                ErrorDetails = errorDetails,
+                CustomData = customData ?? new object()
+            };
+        }
+
+        protected internal static void SendErrorEvent(PlayFabError error)
+        {
+            if (ApiProcessingErrorEventHandler != null)
+            {
+                ApiProcessingErrorEventHandler(error);
+            }
+        }
+
+        protected internal static void SendEvent(object request, object result, ApiProcessingEventType eventType)
+        {
+            if (ApiProcessingEventHandler != null)
+            {
+                ApiProcessingEventHandler(new ApiProcessingEventArgs()
+                {
+                    EventType = eventType,
+                    Request = request,
+                    Result = result
+                });
+            }
+        }
+
+#if PLAYFAB_REQUEST_TIMING
+        protected internal static void SendRequestTiming(RequestTiming rt) {
+            if (ApiRequestTimingEventHandler != null) {
+                ApiRequestTimingEventHandler(rt);
+            }
+        }
+#endif
+        #endregion
+    }
+
+    #region Event Classes
+    public enum ApiProcessingEventType
+    {
+        Pre,
+        Post
+    }
+
+    public class ApiProcessingEventArgs : EventArgs
+    {
+        public ApiProcessingEventType EventType { get; set; }
+        public object Request { get; set; }
+        public object Result { get; set; }
+
+        public T GetRequest<T>()
+        {
+            if (typeof(T) == Request.GetType())
+            {
+                return (T)Request;
+            }
+            return default(T);
```

Comment:
Change your generic to be "where T : PlayFabRequestCommon"
And add a PlayFabRequestCommon base class.
Look for PlayFabResultCommon to see how I do the same thing with results.